### PR TITLE
add support for nvidia gpu access

### PR DESCRIPTION
**Pull request by MondoGao**
Close https://github.com/nextcloud/all-in-one/discussions/4277

Add a new environment to enable Nvidia gpu access when creating containers. It's similar to /dev/dri support. See https://github.com/nextcloud/all-in-one/discussions/1525.

Note: I didn't fully inspect and test this change since I don't have php development environment set up in my machine. Feel free to inherit this pr to enhance its maintainability.

### CODE REVIEW

**Comment by szaimen** [comment]
Hi @MondoGao thanks for your PR! Can you please fix DCO? Wee need that in order to be able to merge it later on. See https://github.com/nextcloud/all-in-one/pull/5132/checks?check_run_id=28844781358

**Comment by MondoGao** [comment]
> Hi @MondoGao thanks for your PR! Can you please fix DCO? Wee need that in order to be able to merge it later on. See https://github.com/nextcloud/all-in-one/pull/5132/checks?check_run_id=28844781358

Sign added :)

**Comment by MondoGao** [comment]
> Hi, I had a fast look at this and I think it would add the capability to all containers that are controlled by AIO. Better would probably be to add this as a capability to [containers-schema.json](https://github.com/nextcloud/all-in-one/blob/main/php/containers-schema.json) and add this only to certain containers via [containers.json](https://github.com/nextcloud/all-in-one/blob/main/php/containers.json)
> 
> Also this is missing some places, e.g. adding documentation on it in the readme. See #1659 as inspiration.
> 
> Additionally, do you know if this is also going to work with AMD and Intel GPUs? In best case we create only one setting that works for all of them.

I believe docker runtime [only supports Nvidia GPU passthrough](https://docs.docker.com/engine/containers/resource_constraints/#gpu). Intel & AMD CPU/GPU's hardware acceleration is exposed through /dev/dri.

Looks like I have to install php dev env to polish this pr, please expect a late response.

**Comment by luzfcb** [comment]
@MondoGao Thank you for starting this effort.

@MondoGao @szaimen 

Besides supporting Nvidia GPU passthrough, there are cases where the container may require the same specific version of the NVIDIA driver installed on the host system. It might be a good idea to create an environment variable for the Nvidia driver version and let the container handle the Nvidia driver setup and configuration logic inside the container. 

That is, Nextcloud AIO could either provide some way for the user to manually specify the version or, if it is not provided by the user, then [have an automatic way to obtain this information](https://github.com/Steam-Headless/docker-steam-headless/blob/860451da74b397385f1b1658545d2bb891aa8e46/overlay/etc/cont-init.d/60-configure_gpu_driver.sh#L18C12-L18C102) and fill an `NVIDIA_DRIVER_VERSION` environment variable.

I have exactly this with docker-steam-headless, and it works great.

This is the driver download and installation script: https://github.com/Steam-Headless/docker-steam-headless/blob/860451da74b397385f1b1658545d2bb891aa8e46/overlay/etc/cont-init.d/60-configure_gpu_driver.sh 


This script creates the X Server configuration files and other related configurations. It probably does not apply to Nextcloud and plugins since they do not use X Server, but I am including it here just for reference.

https://github.com/Steam-Headless/docker-steam-headless/blob/860451da74b397385f1b1658545d2bb891aa8e46/overlay/etc/cont-init.d/70-configure_xorg.sh

**Comment by docjyJ** [comment]
It would be nice to have this feature!

We can have Nextcloud Assistant servers with high performance easily!!!!!!

I agree with @szaimen the configuration should be done in the `containers.json` schema.

Otherwise it looks good to me.

AMD GPU : https://rocm.docs.amd.com/projects/install-on-linux/en/latest/how-to/docker.html

**Comment by gbertolottiKS** [comment]
Just to be sure to have understood this topic correctly: actually there's no way to use a dedicated GPU in Nextcloud AIO, since it is needed a development to make sure it can see (and then use) the graphic card. Is it correct?
This PR should resolve this : once finished, AIO can use "natively" the GPU, right?

Aren't some workaround to be able to use the GPU in the meantime? We are happy to help to test if it's necessary.

Thanks

**Comment by szaimen** [comment]
> Just to be sure to have understood this topic correctly: actually there's no way to use a dedicated GPU in Nextcloud AIO, since it is needed a development to make sure it can see (and then use) the graphic card. Is it correct? This PR should resolve this : once finished, AIO can use "natively" the GPU, right?

Yes

> Aren't some workaround to be able to use the GPU in the meantime? 

Currently no as this PR is unfortunately not even close to being finished. You could try to find someone that takes over finishing this PR to speed up the development.

**Comment by szaimen** [comment]
Also, twig CI seems to fail...

**Comment by docjyJ** [comment]
> > Any news? What's blocking the merger?
> 
> See https://github.com/nextcloud/all-in-one/pull/5132#issuecomment-2296517803

I try to look this

**Comment by docjyJ** [comment]
> Also, twig CI seems to fail...

Fix

**Comment by docjyJ** [comment]
> Hi, I had a fast look at this and I think it would add the capability to all containers that are controlled by AIO. Better would probably be to add this as a capability to [containers-schema.json](https://github.com/nextcloud/all-in-one/blob/main/php/containers-schema.json) and add this only to certain containers via [containers.json](https://github.com/nextcloud/all-in-one/blob/main/php/containers.json)

Done

**Comment by szaimen** [comment]
Thanks @docjyJ! :)

File: compose.yaml [hunk 1]
```diff
@@ -30,6 +30,7 @@ services:
...
       # NEXTCLOUD_ENABLE_DRI_DEVICE: true # This allows to enable the /dev/dri device in the Nextcloud container. ⚠️⚠️⚠️ Warning: this only works if the '/dev/dri' device is present on the host! If it should not exist on your host, don't set this to true as otherwise the Nextcloud container will fail to start! See https://github.com/nextcloud/all-in-one#how-to-enable-hardware-transcoding-for-nextcloud
+      # NEXTCLOUD_NVIDIA_GPU_MODE: 'runtime' # 'runtime' or 'deploy': This allows to enable the [NVIDIA runtime](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html) or [GPU access](https://docs.docker.com/compose/gpu-support/) in the Nextcloud container. Make sure you follow the instructions before setting this value. If you're using WSL2 and want to use the NVIDIA runtime, please follow the instructions to [install the NVIDIA Container Toolkit meta-version in WSL](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#cuda-support-for-wsl-2).
```
Comment (szaimen) [inline compose.yaml:33]: Can we maybe rename it to a different name, and for example simply name it `NEXTCLOUD_ENABLE_GPU: nvidia` for example? We could then also use this setting for amd or intel later on...

File: php/containers-schema.json [hunk 2]
```diff
@@ -160,6 +160,9 @@
 							"pattern": "^/dev/[a-z]+$"
 						}
 					},
+					"nvidia": {
```
Comment (szaimen) [inline php/containers-schema.json:163]: can we maybe name this nextcloud_enable_gpu?

File: php/public/index.php [hunk 3]
```diff
@@ -125,6 +125,8 @@
         'nextcloud_max_time' => $configurationManager->GetNextcloudMaxTime(),
         'nextcloud_memory_limit' => $configurationManager->GetNextcloudMemoryLimit(),
         'is_dri_device_enabled' => $configurationManager->isDriDeviceEnabled(),
+        'is_nvidia_runtime_enabled' => $configurationManager->isNvidiaRuntimeEnabled(),
+        'is_nvidia_gpu_deploy_enabled' => $configurationManager->isNvidiaDeployEnabled(),
```
Comment (szaimen) [inline php/public/index.php:129]: can we only have one setting for this?

**Review by szaimen — CHANGES_REQUESTED** [review]
Thanks @docjyJ :)
A few comments below...

**Comment by szaimen** [comment]
Also, what about talk-recording and fulltextsearch?

File: compose.yaml (hunk 1 above)
Comment (szaimen) [reply compose.yaml:33]: > Okay... Leave future problems for the future.

sounds good

Comment (docjyJ) [reply compose.yaml:33]: See https://github.com/mviereck/x11docker/wiki/Hardware-acceleration

File: readme.md [hunk 4]
```diff
@@ -765,11 +765,33 @@ You can do so by adding `--env NEXTCLOUD_ADDITIONAL_PHP_EXTENSIONS="imagick exte
...
+
+> [!WARNING]  
+> This feature is in beta. Since the proprietary, we haven't a lot of user using proprietary drivers, we can't guarantee the stability of this feature.
+> Your feedback is welcome.
+
+This methode use the [Nvidia Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/index.html) with the nvidia runtime.
+
+To enable it, use `--env NEXTCLOUD_GPU_MODE=nvidia-runtime` as driver mode.
 
-The [memories app](https://apps.nextcloud.com/apps/memories) allows to enable hardware transcoding for videos. In order to use that, you need to add `--env NEXTCLOUD_ENABLE_DRI_DEVICE=true` to the docker run command of the mastercontainer (but before the last line `nextcloud/all-in-one:latest`! If it was started already, you will need to stop the mastercontainer, remove it (no data will be lost) and recreate it using the docker run command that you initially used) which will mount the `/dev/dri` device into the container. There is now a community container which allows to easily add the transcoding container of Memories to AIO: https://github.com/nextcloud/all-in-one/tree/main/community-containers/memories
```
Comment (szaimen) [inline readme.md:772]: Sorry @docjyJ there was a misunderstanding: I would keep the existing setting `NEXTCLOUD_ENABLE_DRI_DEVICE`. My request was onky about the new setting to have one setting for all external gpus, no whether they are nvidia, amd or intel based...

File: compose.yaml (hunk 1 above)
Comment (docjyJ) [reply compose.yaml]: c64908e

File: readme.md (hunk 4 above)
Comment (szaimen) [reply readme.md:772]: The dri device is the hardware encoding device present in cpus. This should definitely be possibly to enable it in parallel to gpus - thus an additional switch.

Comment (docjyJ) [reply readme.md:772]: 0fdd137

**Comment by docjyJ** [comment]
It seems that it does not use the CPU but the integrated graphics components.

> The Direct Rendering Infrastructure (DRI) is the framework comprising the modern [Linux](https://en.wikipedia.org/wiki/Linux) graphics stack which allows unprivileged user-space programs to issue commands to [graphics hardware](https://en.wikipedia.org/wiki/Graphics_hardware) without conflicting with other programs.

https://en.wikipedia.org/wiki/Direct_Rendering_Infrastructure

> Intel GMA, HD Graphics, Iris. See [Intel’s Website](https://www.intel.com/content/www/us/en/developer/topic-technology/open/overview.html)

https://docs.mesa3d.org/systems.html

**Comment by szaimen** [comment]
> It seems that it does not use the CPU but the integrated graphics components.

Yes, you are right :D

**Comment by szaimen** [comment]
@docjyJ is it clear what I intend?

- keep NEXTCLOUD_ENABLE_DRI_DEVICE
- have one additional setting for all external gpus like e.g. `NEXTCLOUD_EXTERNAL_GPU_MODE: nvidia-runtime`

Does this make sense? :)

**Comment by docjyJ** [comment]
> @docjyJ is it clear what I intend?
> 
> - keep `NEXTCLOUD_ENABLE_DRI_DEVICE`
> - have one additional setting for all external gpus like e.g. `NEXTCLOUD_EXTERNAL_GPU_MODE: nvidia-runtime`
> 
> Does this make sense? :)

See last commit

**Comment by docjyJ** [comment]
> btw, we should also add this setting to https://github.com/nextcloud/all-in-one/tree/main/community-containers/facerecognition and https://github.com/nextcloud/all-in-one/tree/main/community-containers/local-ai :)

An other pr ?

**Comment by szaimen** [comment]
> > @docjyJ is it clear what I intend?
> > 
> > * keep `NEXTCLOUD_ENABLE_DRI_DEVICE`
> > * have one additional setting for all external gpus like e.g. `NEXTCLOUD_EXTERNAL_GPU_MODE: nvidia-runtime`
> > 
> > Does this make sense? :)
> 
> See last commit

But you are back to `NEXTCLOUD_NVIDIA_GPU_MODE` now... Cant we name it NEXTCLOUD_EXTERNAL_GPU_MODE for example? Same for enable_nvidia_gpu in containers-schema... Can we name it `enable_external_gpu` for example?

**Comment by docjyJ** [comment]
> But you are back to `NEXTCLOUD_NVIDIA_GPU_MODE` now... Cant we name it `NEXTCLOUD_EXTERNAL_GPU_MODE` for example? Same for `enable_nvidia_gpu` in containers-schema... Can we name it `enable_external_gpu` for example?

I disagree, as I don't know of any other brand that doesn't use `/dev/dri`.

This feature is specific to the Nvidia proprietary driver.

Most open-source drivers use `/dev/dri`. This also includes the new Nvidia open-source driver.

Also, since Nvidia maintains an open-source version of its drivers, I realize that this feature may be useless...

**Comment by docjyJ** [comment]
Some reference:
https://de.wikipedia.org/wiki/Nouveau

https://nouveau.freedesktop.org/
https://dri.freedesktop.org/

https://docs.mesa3d.org/systems.html

**Review by docjyJ — COMMENTED** [review]
LGTM 🎉

**Comment by docjyJ** [comment]
Thank you @MondoGao

**Review by szaimen — APPROVED** [review]
LGTM. Thanks a lot! 😊

**Comment by szaimen** [comment]
This is now released with v10.2.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Comment by luzfcb** [comment]
@szaimen the `local-ai` community container is unable to start with Nextcloud AIO configured with the  `NEXTCLOUD_ENABLE_DRI_DEVICE=true`, `ENABLE_NVIDIA_GPU=true`, `AIO_COMMUNITY_CONTAINERS=local-ai` environment variables.

This is the logs that I have in the AIO container:

<details>
  <summary>Click me to see the logs</summary>
  
```bash


Initial startup of Nextcloud All-in-One complete!
You should be able to open the Nextcloud AIO Interface now on port 8080 of this server!
E.g. https://internal.ip.of.this.server:8080
⚠️ Important: do always use an ip-address if you access this port and not a domain as HSTS might block access to it later!

If your server has port 80 and 8443 open and you point a domain to your server, you can get a valid certificate automatically by opening the Nextcloud AIO Interface via:
https://your-domain-that-points-to-this-server.tld:8443
NOTICE: PHP message: Slim Application Error
Type: Exception
Code: 0
Message: Could not start container nextcloud-aio-local-ai: Server error: `POST http://127.0.0.1/v1.41/containers/nextcloud-aio-local-ai/start` resulted in a `500 Internal Server Error` response:
{"message":"failed to create task for container: failed to create shim task: OCI runtime create failed: runc create fail (truncated...)
File: /var/www/docker-aio/php/src/Docker/DockerActionManager.php
Line: 170
Trace: #0 /var/www/docker-aio/php/src/Controller/DockerController.php(59): AIO\Docker\DockerActionManager->StartContainer(Object(AIO\Container\Container))
#1 /var/www/docker-aio/php/src/Controller/DockerController.php(26): AIO\Controller\DockerController->PerformRecursiveContainerStart('nextcloud-aio-l...', true)
#2 /var/www/docker-aio/php/src/Controller/DockerController.php(209): AIO\Controller\DockerController->PerformRecursiveContainerStart('nextcloud-aio-a...', true)
#3 /var/www/docker-aio/php/src/Controller/DockerController.php(189): AIO\Controller\DockerController->startTopContainer(true)
#4 /var/www/docker-aio/php/vendor/slim/slim/Slim/Handlers/Strategies/RequestResponse.php(38): AIO\Controller\DockerController->StartContainer(Object(GuzzleHttp\Psr7\ServerRequest), Object(GuzzleHttp\Psr7\Response), Array)
#5 /var/www/docker-aio/php/vendor/slim/slim/Slim/Routing/Route.php(363): Slim\Handlers\Strategies\RequestResponse->__invoke(Array, Object(GuzzleHttp\Psr7\ServerRequest), Object(GuzzleHttp\Psr7\Response), Array)
#6 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(73): Slim\Routing\Route->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#7 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(73): Slim\MiddlewareDispatcher->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#8 /var/www/docker-aio/php/vendor/slim/slim/Slim/Routing/Route.php(321): Slim\MiddlewareDispatcher->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#9 /var/www/docker-aio/php/vendor/slim/slim/Slim/Routing/RouteRunner.php(74): Slim\Routing\Route->run(Object(GuzzleHttp\Psr7\ServerRequest))
#10 /var/www/docker-aio/php/vendor/slim/csrf/src/Guard.php(482): Slim\Routing\RouteRunner->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#11 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(177): Slim\Csrf\Guard->process(Object(GuzzleHttp\Psr7\ServerRequest), Object(Slim\Routing\RouteRunner))
#12 /var/www/docker-aio/php/vendor/slim/twig-view/src/TwigMiddleware.php(117): Psr\Http\Server\RequestHandlerInterface@anonymous->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#13 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(129): Slim\Views\TwigMiddleware->process(Object(GuzzleHttp\Psr7\ServerRequest), Object(Psr\Http\Server\RequestHandlerInterface@anonymous))
#14 /var/www/docker-aio/php/src/Middleware/AuthMiddleware.php(36): Psr\Http\Server\RequestHandlerInterface@anonymous->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#15 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(280): AIO\Middleware\AuthMiddleware->__invoke(Object(GuzzleHttp\Psr7\ServerRequest), Object(Psr\Http\Server\RequestHandlerInterface@anonymous))
#16 /var/www/docker-aio/php/vendor/slim/slim/Slim/Middleware/ErrorMiddleware.php(77): Psr\Http\Server\RequestHandlerInterface@anonymous->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#17 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(129): Slim\Middleware\ErrorMiddleware->process(Object(GuzzleHttp\Psr7\ServerRequest), Object(Psr\Http\Server\RequestHandlerInterface@anonymous))
#18 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(73): Psr\Http\Server\RequestHandlerInterface@anonymous->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#19 /var/www/docker-aio/php/vendor/slim/slim/Slim/App.php(209): Slim\MiddlewareDispatcher->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#20 /var/www/docker-aio/php/vendor/slim/slim/Slim/App.php(193): Slim\App->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#21 /var/www/docker-aio/php/public/index.php(189): Slim\App->run()
#22 {main}
Tips: To display error details in HTTP response set "displayErrorDetails" to true in the ErrorHandler constructor.
NOTICE: PHP message: Slim Application Error
Type: Exception
Code: 0
Message: Could not start container nextcloud-aio-local-ai: Server error: `POST http://127.0.0.1/v1.41/containers/nextcloud-aio-local-ai/start` resulted in a `500 Internal Server Error` response:
{"message":"failed to create task for container: failed to create shim task: OCI runtime create failed: runc create fail (truncated...)
File: /var/www/docker-aio/php/src/Docker/DockerActionManager.php
Line: 170
Trace: #0 /var/www/docker-aio/php/src/Controller/DockerController.php(59): AIO\Docker\DockerActionManager->StartContainer(Object(AIO\Container\Container))
#1 /var/www/docker-aio/php/src/Controller/DockerController.php(26): AIO\Controller\DockerController->PerformRecursiveContainerStart('nextcloud-aio-l...', true)
#2 /var/www/docker-aio/php/src/Controller/DockerController.php(209): AIO\Controller\DockerController->PerformRecursiveContainerStart('nextcloud-aio-a...', true)
#3 /var/www/docker-aio/php/src/Controller/DockerController.php(189): AIO\Controller\DockerController->startTopContainer(true)
#4 /var/www/docker-aio/php/vendor/slim/slim/Slim/Handlers/Strategies/RequestResponse.php(38): AIO\Controller\DockerController->StartContainer(Object(GuzzleHttp\Psr7\ServerRequest), Object(GuzzleHttp\Psr7\Response), Array)
#5 /var/www/docker-aio/php/vendor/slim/slim/Slim/Routing/Route.php(363): Slim\Handlers\Strategies\RequestResponse->__invoke(Array, Object(GuzzleHttp\Psr7\ServerRequest), Object(GuzzleHttp\Psr7\Response), Array)
#6 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(73): Slim\Routing\Route->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#7 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(73): Slim\MiddlewareDispatcher->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#8 /var/www/docker-aio/php/vendor/slim/slim/Slim/Routing/Route.php(321): Slim\MiddlewareDispatcher->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#9 /var/www/docker-aio/php/vendor/slim/slim/Slim/Routing/RouteRunner.php(74): Slim\Routing\Route->run(Object(GuzzleHttp\Psr7\ServerRequest))
#10 /var/www/docker-aio/php/vendor/slim/csrf/src/Guard.php(482): Slim\Routing\RouteRunner->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#11 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(177): Slim\Csrf\Guard->process(Object(GuzzleHttp\Psr7\ServerRequest), Object(Slim\Routing\RouteRunner))
#12 /var/www/docker-aio/php/vendor/slim/twig-view/src/TwigMiddleware.php(117): Psr\Http\Server\RequestHandlerInterface@anonymous->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#13 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(129): Slim\Views\TwigMiddleware->process(Object(GuzzleHttp\Psr7\ServerRequest), Object(Psr\Http\Server\RequestHandlerInterface@anonymous))
#14 /var/www/docker-aio/php/src/Middleware/AuthMiddleware.php(36): Psr\Http\Server\RequestHandlerInterface@anonymous->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#15 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(280): AIO\Middleware\AuthMiddleware->__invoke(Object(GuzzleHttp\Psr7\ServerRequest), Object(Psr\Http\Server\RequestHandlerInterface@anonymous))
#16 /var/www/docker-aio/php/vendor/slim/slim/Slim/Middleware/ErrorMiddleware.php(77): Psr\Http\Server\RequestHandlerInterface@anonymous->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#17 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(129): Slim\Middleware\ErrorMiddleware->process(Object(GuzzleHttp\Psr7\ServerRequest), Object(Psr\Http\Server\RequestHandlerInterface@anonymous))
#18 /var/www/docker-aio/php/vendor/slim/slim/Slim/MiddlewareDispatcher.php(73): Psr\Http\Server\RequestHandlerInterface@anonymous->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#19 /var/www/docker-aio/php/vendor/slim/slim/Slim/App.php(209): Slim\MiddlewareDispatcher->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#20 /var/www/docker-aio/php/vendor/slim/slim/Slim/App.php(193): Slim\App->handle(Object(GuzzleHttp\Psr7\ServerRequest))
#21 /var/www/docker-aio/php/public/index.php(189): Slim\App->run()
#22 {main}
Tips: To display error details in HTTP response set "displayErrorDetails" to true in the ErrorHandler constructor.
```
</details>

I think the reason is probably because the docker image used as a base for aio-local-ai is quay.io/go-skynet/local-ai:v2.24.2-aio-cpu , but when `ENABLE_NVIDIA_GPU=true` is used, it should be one of the images below:

quay.io/go-skynet/local-ai:v2.24.2-aio-gpu-nvidia-cuda-12
quay.io/go-skynet/local-ai:v2.24.2-aio-gpu-nvidia-cuda-11

Additionally, I saw that there are options for Intel GPU or something like this

quay.io/go-skynet/local-ai:v2.24.2-aio-gpu-intel-f16
quay.io/go-skynet/local-ai:v2.24.2-aio-gpu-intel-f32
quay.io/go-skynet/local-ai:v2.24.2-aio-gpu-hipblas

**Comment by szaimen** [comment]
Hm... This specific error should not happen, it should still be able to start the container... Did you install the nvidia drivers correctly like mentioned im the readme?

**Comment by luzfcb** [comment]
> Did you install the nvidia drivers correctly like mentioned im the readme?

Yes, since I can successfully run the Sample Workload with Docker

```
sudo docker run --rm --runtime=nvidia --gpus all ubuntu nvidia-smi
```

image

Note:

I'm using Unraid OS 6.12.3 and have the Nvidia Driver plugin by ich777 and the Nvidia proprietary driver v565.57.01 , and I am able to play games on Steam via docker container https://github.com/Steam-Headless/docker-steam-headless


Anyway, the main problem about the error on the local-ai community container is because it somehow block the all other nextcloud containers to start by Nextcloud AIO.

**Comment by szaimen** [comment]
So if you remove local-ai from the community containers, nextcloud starts correctly?

**Comment by luzfcb** [comment]
> So if you remove local-ai from the community containers, nextcloud starts correctly?

Yes, and also if I keep local-ai on the community containers and set ENABLE_NVIDIA_GPU=false the nextcloud also starts correctly

**Comment by szaimen** [comment]
Okay this is weird. I guess we need to remove the setting for local-ai then. Doe the other containers that we added start correctly?

**Comment by luzfcb** [comment]
>  I guess we need to remove the setting for local-ai then

Until `aio-local-ai` provides a new image using `quay.io/go-skynet/local-ai:v2.24.2-aio-gpu-nvidia-cuda-1`2 as base instead of `quay.io/go-skynet/local-ai:v2.24.2-aio-cpu` , probably yes

> Doe the other containers that we added start correctly?

`facerecognition` and `memories` start correctly. 

Since I don't use `plex` and `jellyfin`, and I can't confirm if they work or not.

I'm also not sure if `facerecognition` and `memories` recognize the nvidia GPU immediately without any extra configuration.

I'll try to figure that out over the weekend.

**Comment by luzfcb** [comment]
>  I guess we need to remove the setting for local-ai then

Until `aio-local-ai` provides a new image using `quay.io/go-skynet/local-ai:v2.24.2-aio-gpu-nvidia-cuda-12` as base instead of `quay.io/go-skynet/local-ai:v2.24.2-aio-cpu` , and community containers feature implements a way to change the image or tag defined on https://github.com/nextcloud/all-in-one/blob/109b9dc019ebb499a9571f8cf3129e6e26e1942a/community-containers/local-ai/local-ai.json#L7-L8 based on the existence of the `ENABLE_NVIDIA_GPU=true`, probably yes

> Doe the other containers that we added start correctly?

`facerecognition` and `memories` start correctly. 

Since I don't use `plex` and `jellyfin`, and I can't confirm if they work or not.

I'm also not sure if `facerecognition` and `memories` recognize the nvidia GPU immediately without any extra configuration.

I'll try to figure that out over the weekend.

**Comment by luzfcb** [comment]
>>  I guess we need to remove the setting for local-ai then

> Until `aio-local-ai` provides a new image using `quay.io/go-skynet/local-ai:v2.24.2-aio-gpu-nvidia-cuda-12` as base instead of `quay.io/go-skynet/local-ai:v2.24.2-aio-cpu` , and community containers feature implements a way to change the image or tag defined on `https://github.com/nextcloud/all-in-one/blob/109b9dc019ebb499a9571f8cf3129e6e26e1942a/community-containers/local-ai/local-ai.json#L7-L8` based on the existence of the `ENABLE_NVIDIA_GPU=true`, probably yes


A quick and simple solution is probably to create a new community container named `local-ai-nvidia` that uses a possible new `aio-local-ai` nvidia based image.

**Comment by deftmartian** [comment]
> > I guess we need to remove the setting for local-ai then
> 
> Until `aio-local-ai` provides a new image using `quay.io/go-skynet/local-ai:v2.24.2-aio-gpu-nvidia-cuda-12` as base instead of `quay.io/go-skynet/local-ai:v2.24.2-aio-cpu` , and community containers feature implements a way to change the image or tag defined on
> 
> https://github.com/nextcloud/all-in-one/blob/109b9dc019ebb499a9571f8cf3129e6e26e1942a/community-containers/local-ai/local-ai.json#L7-L8
> based on the existence of the `ENABLE_NVIDIA_GPU=true`, probably yes
> 
> > Doe the other containers that we added start correctly?
> 
> `facerecognition` and `memories` start correctly.
> 
> Since I don't use `plex` and `jellyfin`, and I can't confirm if they work or not.
> 
> I'm also not sure if `facerecognition` and `memories` recognize the nvidia GPU immediately without any extra configuration.
> 
> I'll try to figure that out over the weekend.

facerecognition and memories don't seem to see the gpu immediately.   I tried `ENABLE_NVIDIA_GPU: true` alone as well as with ` NEXTCLOUD_ENABLE_DRI_DEVICE: true` in my compose (wasn't clear if I should have both options enabled).   Fully removed master container and rebuilt. 

facerecognition is running through 70k photos, and I don't see any gpu activity in nvtop.  Tried resetting facerecognition and restarting for all configurations.

I currently use the external go-vod container for memories.  Disabling this but leaving transcoding on didn't work.

My nvidia runtime is configured correctly.  I use seperate containers for ai and jellyfin, and they both use the nvidia runtime with no issues.  

I'm not sure where to look for logs, but I am happy to help if given some direction.

**Comment by szaimen** [comment]
> Okay this is weird. I guess we need to remove the setting for local-ai then.

This is now fixed with v10.3.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Pull request merged**